Library/OcAfterBootCompatLib/RelocationBlock.c

```diff
@@ -290,9 +332,43 @@ AppleRelocationRebase (
     }
   }
 
+  //
+  // On macOS 10.4 and 10.5, EfiBoot has already set the system table address to the correct virtual one.
+  // The memory map also contains additional trashed entries that must be removed, this seems to occur on Macs as well.
+  //
+  // On macOS 10.6 and newer, we need to adjust it like the others.
+  //
+  if (!BootCompat->KernelState.RelocationBlockLegacy) {
+    *BA->SystemTableP -= RelocDiff;
+  } else {
+    MemoryMapSize  = *BA->MemoryMapSize;
+    DescriptorSize = *BA->MemoryMapDescriptorSize;
+    MemoryMap      = (EFI_MEMORY_DESCRIPTOR *)(UINTN)*BA->MemoryMap;
+
+    Desc               = MemoryMap;
+    DescLargestAddress = Desc->PhysicalStart;
+    NumEntries         = MemoryMapSize / DescriptorSize;
+
+    //
+    // Locate end of valid memory map. It is assumed that the entries are in order from
+    // lowest to highest address, performed by RebuildAppleMemoryMap booter quirk.
+    //
+    for (Index = 0; Index < NumEntries; ++Index) {
+      if (Desc->PhysicalStart >= DescLargestAddress) {
+        DescLargestAddress = Desc->PhysicalStart;
```

Comment:
The `DescLargestAddress` name and the control flow format make it look like it looks for a maximum, but that's not the case. I would invert the if-condition (move else branch up), move this outside the branch (it will `break`), and rename the variable to `PrevDescAddress`. I'm not sure this is safe though, some firmwares do this OOTB. What is the exact issue?